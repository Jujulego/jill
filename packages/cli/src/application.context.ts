import { createContext, FC, useContext } from 'react';
import yargs from 'yargs';

// Test
export type Args<A> = yargs.ArgumentsCamelCase<A> & { '--': readonly (string | number)[] };
export type Builder<A> = (yargs: yargs.Argv) => yargs.Argv<A>;
export type BuilderWrapper<A> = <Args>(yargs: yargs.Argv<Args>) => yargs.Argv<Omit<Args, keyof A> & A>;

export interface GlobalArgs {
  plugins: string[];
  verbose: number;
}

export interface Command<A> {
  id: string;
  name: string | readonly [string, ...string[]];
  description: string;
  builder: Builder<A>;
}

export interface ApplicationContextState {
  args: Args<unknown>;
  command?: Command<unknown>;
}

export type CommandComponent<A, P> = FC<P> & {
  command: Command<A>;
}

export type UseArgsHook<A> = () => Args<A & GlobalArgs>;

// Context
export const applicationDefaultState: ApplicationContextState = {
  args: {
    '$0': '',
    '_': [],
    '--': []
  }
};

export const ApplicationContext = createContext(applicationDefaultState);

// Hooks
export function useArgs<A>(): Args<A & GlobalArgs> {
  const { args } = useContext(ApplicationContext);
  return args as Args<A & GlobalArgs>;
}
