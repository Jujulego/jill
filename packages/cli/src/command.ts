import { logger } from '@jujulego/jill-core';
import yargs from 'yargs';

import { transport } from './logger';

// Export
export type Awaitable<T> = T | PromiseLike<T>;
export type Arguments<A> = yargs.Arguments<A> & { '--': readonly (string | number)[] };
export type Builder<T, A = T> = (yargs: yargs.Argv<T>) => yargs.Argv<A>;

// Command
export abstract class Command<A = unknown> {
  // Attributes
  readonly logger = logger;
  readonly spinner = transport.spinner;

  // Methods
  protected abstract define<T, U>(builder: Builder<T, U>): Builder<T, U & A>;
  protected abstract run(args: Arguments<A>): Awaitable<number | void>;

  setup<T>(yargs: yargs.Argv<T>): yargs.Argv<T> {
    yargs.command<A>(
      this.name,
      this.description,
      (y) => this.define(_ => _)(y),
      (a) => this._wrapper(a as Arguments<A>)
    );

    return yargs;
  }

  private async _wrapper(args: Arguments<A>): Promise<void> {
    try {
      const exit = await this.run(args);
      process.exit(exit ?? 0);
    } catch (err) {
      this.spinner.fail(err.message);
      process.exit(1);
    }
  }

  log(msg: string): void {
    if (this.spinner.isSpinning) {
      this.spinner.clear();
    }

    console.log(msg);
  }

  // Properties
  abstract get name(): string | readonly string[];
  abstract get description(): string;
}