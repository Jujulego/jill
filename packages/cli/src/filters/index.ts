export { AffectedFilter } from './affected';
export { Filter } from './filter';