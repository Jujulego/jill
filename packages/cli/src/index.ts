export { buildCommand } from './commands/build';
export { eachCommand } from './commands/each';
export { infoCommand } from './commands/info';
export { logger, OraLogger } from './logger';
