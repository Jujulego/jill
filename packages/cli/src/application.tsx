import { Children, FC, isValidElement, ReactElement, useEffect, useState } from 'react';
import Spinner from 'ink-spinner';
import { Text } from 'ink';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

import { CommandComponent } from './command.hoc';
import { ApplicationContext, ApplicationContextState, applicationDefaultState, Args } from './application.context';

// Types
export interface ApplicationProps {
  name: string;
}

// Constants
const parser = yargs(hideBin(process.argv))
  .parserConfiguration({
    'populate--': true,
  });

const commands = new Map<string, ReactElement>();

// Component
export const Application: FC<ApplicationProps> = ({ name, children }) => {
  // State
  const [state, setState] = useState<ApplicationContextState>(applicationDefaultState);

  // Effects
  useEffect(() => {
    // Config yargs
    parser.scriptName(name);

    // Define commands
    Children.map(children, (child) => {
      if (!isValidElement(child)) {
        return;
      }

      const { command } = child.type as CommandComponent<unknown, unknown>;
      commands.set(command.id, child);
      parser.command(
        command.name,
        command.description,
        (y) => command.builder(y),
        (args) => {
          setState((old) => ({
            ...old,
            args: args as Args<unknown>,
            command
          }));
        }
      );
    });

    // Parse to run command
    parser.strictCommands()
      .help()
      .parse();
  }, [name, children]);

  // Render
  if (!state.command) {
    return (
      <Text>
        <Spinner />
        {' '} Loading { name } ...
      </Text>
    );
  }

  return (
    <ApplicationContext.Provider value={state}>
      { commands.get(state.command.id) }
    </ApplicationContext.Provider>
  );
};
