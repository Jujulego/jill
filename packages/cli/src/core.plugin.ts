import { Plugin } from '@jujulego/jill-common';

import { EachCommand } from './commands/each.command';
import { RunCommand } from './commands/run.command';

// Plugin
export const corePlugin = Plugin.createPlugin('core', [
  RunCommand,
  EachCommand,
]);
