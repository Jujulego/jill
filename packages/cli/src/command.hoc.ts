import { FC } from 'react';

import { Args, Command, useArgs } from './application.context';

// Types
export type CommandMetadata<A> = Omit<Command<A>, 'id'>

export type CommandComponent<A, P> = FC<P> & {
  command: Command<A>;
}

export interface CommandUtils<A> {
  useArgs: () => Args<A>;
  wrapper: <P>(Component: FC<P>) => CommandComponent<A, P>;
}

// Generate HOC & Hook for command components
export function command<A>(command: CommandMetadata<A>): CommandUtils<A> {
  return {
    useArgs: () => useArgs<A>(),
    wrapper: <P>(Component: FC<P>): CommandComponent<A, P> => {
      Component.displayName = `command(${Component.displayName || Component.name})`;

      return Object.assign(Component, {
        command: {
          ...command,
          id: typeof command.name === 'string' ? command.name : command.name[0],
        }
      });
    }
  };
}
