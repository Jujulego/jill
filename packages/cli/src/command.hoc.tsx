import { FC, useContext } from 'react';
import yargs from 'yargs';
import { ArgsContext } from './application';

// Types
export type Arguments<A> = yargs.Arguments<A> & { '--': readonly (string | number)[] };
export type Builder<A> = (yargs: yargs.Argv) => yargs.Argv<A>;

export interface CommandMetadata<A> {
  name: string | readonly [string, ...string[]];
  description: string;
  define: Builder<A>;
}

export type CommandComponent<A, P = Record<string, unknown>> = FC<P> & {
  command: CommandMetadata<A> & { id: string };
}

// HOC
export function command<A, P = Record<string, unknown>>(command: CommandMetadata<A>, Component: FC<P & Arguments<A>>): CommandComponent<A, P> {
  const Wrapper: FC<P> = (props) => {
    const args = useContext(ArgsContext) as Arguments<A>;

    return <Component {...props} {...args} />;
  };

  Wrapper.displayName = `command(${Component.name || Component.displayName})`;

  return Object.assign(Wrapper, {
    command: {
      ...command,
      id: typeof command.name === 'string' ? command.name : command.name[0]
    }
  });
}
