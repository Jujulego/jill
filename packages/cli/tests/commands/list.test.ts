import { Project, Workspace } from '@jujulego/jill-core';
import chalk from 'chalk';

import { ListArgs, listCommand, logger } from '../../src';

// Setup
jest.mock('../../src/logger');

chalk.level = 1;

const defaults: ListArgs = {
  affected: undefined,
  private: undefined,
  'with-script': undefined,

  attrs: undefined,
  headers: undefined,
  long: false,
  json: false,
};

let project: Project;
let screen: string;

beforeEach(() => {
  project = new Project('.');
  screen = '';

  // Mocks
  jest.restoreAllMocks();

  jest.spyOn(console, 'log').mockImplementation((message) => screen += message + '\n');
});

// Tests
describe('jill list', () => {
  // Setup
  let workspaces: Workspace[];

  beforeEach(() => {
    workspaces = [
      new Workspace('./wks-1', { name: 'wks-1', private: true, version: '1.0.0' }, project),
      new Workspace('./wks-2', { name: 'wks-2', version: '1.0.0', scripts: { test: 'test' } }, project),
      new Workspace('./wks-3', { name: 'wks-3', version: '1.0.0' }, project),
    ];

    jest.spyOn(project, 'workspaces').mockImplementation(async function* () {
      for (const wks of workspaces) yield wks;
    });

    for (const wks of workspaces) {
      jest.spyOn(wks, 'isAffected').mockResolvedValue(wks.name === 'wks-2');
    }
  });

  // Defaults
  it('should print list of all workspaces', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false }))
      .resolves.toBe(0);

    // Checks
    expect(logger.spin).toHaveBeenCalledWith('Loading project');
    expect(project.workspaces).toHaveBeenCalled();
    expect(logger.stop).toHaveBeenCalled();
    expect(screen).toEqual('wks-1\nwks-2\nwks-3\n');
  });

  // Filters
  it('should print only private workspaces (--private)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false, private: true }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toEqual('wks-1\n');
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  it('should print only public workspaces (--no-private)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false, private: false }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toEqual('wks-2\nwks-3\n');
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  it('should print only affected workspaces (--affected test)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false, affected: 'test' }))
      .resolves.toBeUndefined();

    // Checks
    for (const wks of workspaces) {
      expect(wks.isAffected).toHaveBeenCalledWith('test');
    }

    expect(screen).toEqual('wks-2\n');
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  it('should print only workspaces with \'test\' script (--with-script test)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false, 'with-script': 'test' }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toEqual('wks-2\n');
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  // Formats
  it('should print list with headers (--headers)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: false, headers: true }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toMatchSnapshot();
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  it('should print long list of all workspaces (--long)', async () => {
    // Call
    await expect(listCommand(project, { ...defaults, long: true, json: false }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toMatchSnapshot();
    expect(process.exit).toHaveBeenCalledWith(0);
  });

  it('should print json array of all workspaces (--json)', async () => {
    for (const wks of workspaces) {
      jest.spyOn(wks, 'cwd', 'get').mockReturnValue(`/full/path/to/${wks.name}`);
    }

    // Call
    await expect(listCommand(project, { ...defaults, long: false, json: true }))
      .resolves.toBeUndefined();

    // Checks
    expect(screen).toMatchSnapshot();
    expect(process.exit).toHaveBeenCalledWith(0);
  });
});
