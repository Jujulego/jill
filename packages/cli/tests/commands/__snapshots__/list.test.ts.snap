// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jill list should print json array of all workspaces (--json) 1`] = `
"[{\\"name\\":\\"wks-1\\",\\"version\\":\\"1.0.0\\",\\"root\\":\\"/full/path/to/wks-1\\"},{\\"name\\":\\"wks-2\\",\\"version\\":\\"1.0.0\\",\\"root\\":\\"/full/path/to/wks-2\\"},{\\"name\\":\\"wks-3\\",\\"version\\":\\"1.0.0\\",\\"root\\":\\"/full/path/to/wks-3\\"}]
"
`;

exports[`jill list should print list with headers (--headers) 1`] = `
"[1mName [22m
wks-1
wks-2
wks-3
"
`;

exports[`jill list should print long list of all workspaces (--long) 1`] = `
"[1mName [22m  [1mVersion[22m  [1mRoot [22m
wks-1  1.0.0    wks-1
wks-2  1.0.0    wks-2
wks-3  1.0.0    wks-3
"
`;
